digest/src/variable.rs

```diff
@@ -0,0 +1,104 @@
+//! Variable-sized output digest support
+
+use crate::{InvalidOutputSize, Reset};
+
+#[cfg(feature = "alloc")]
+use alloc::boxed::Box;
+
+/// Trait for returning digest result with the variable size
+pub trait VariableOutput: Sized {
+    /// Create new hasher instance with the given output size.
+    ///
+    /// It will return `Err(InvalidOutputSize)` in case if hasher can not return
+    /// specified output size. It will always return an error if output size
+    /// equals to zero.
+    fn new(output_size: usize) -> Result<Self, InvalidOutputSize>;
+
+    /// Get output size of the hasher instance provided to the `new` method
+    fn output_size(&self) -> usize;
+
+    /// Retrieve result via closure and consume hasher.
+    ///
+    /// Closure is guaranteed to be called, length of the buffer passed to it
+    /// will be equal to `output_size`.
+    fn finalize_variable(self, f: impl FnOnce(&[u8]));
+
+    /// Retrieve result via closure and reset the hasher state.
+    ///
+    /// Closure is guaranteed to be called, length of the buffer passed to it
+    /// will be equal to `output_size`.
+    fn finalize_variable_reset(&mut self, f: impl FnOnce(&[u8]));
+
+    /// Retrieve result into a boxed slice and consume hasher.
+    ///
+    /// `Box<[u8]>` is used instead of `Vec<u8>` to save stack space, since
+    /// they have size of 2 and 3 words respectively.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    fn finalize_boxed(self) -> Box<[u8]> {
+        let n = self.output_size();
+        let mut buf = vec![0u8; n].into_boxed_slice();
+        self.finalize_variable(|res| buf.copy_from_slice(res));
+        buf
+    }
+
+    /// Retrieve result into a boxed slice and reset hasher state.
+    ///
+    /// `Box<[u8]>` is used instead of `Vec<u8>` to save stack space, since
+    /// they have size of 2 and 3 words respectively.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    fn finalize_boxed_reset(&mut self) -> Box<[u8]> {
+        let n = self.output_size();
+        let mut buf = vec![0u8; n].into_boxed_slice();
```

Comment:
Hm, don't we have to explicitly import the `vec` macro? Does CI test build with an enabled `alloc` feature?

Reply:
> Hm, don't we have to explicitly import the vec macro?

It's being imported from the alloc crate:

https://github.com/RustCrypto/traits/blob/master/digest/src/lib.rs#L23

>  Does CI test build with an enabled `alloc` feature?

Yes, it tests each feature individually as well as `--all-features` (and the build with `--no-default-features` on Cortex-M and WASM)

https://github.com/RustCrypto/traits/blob/master/.github/workflows/digest.yml#L55

Reply:
Ah, great. Haven't used `#[macro_use]` in a long while. :)

Reply:
It might be possible to import it as `use alloc::vec` instead but the nice thing about the `#[macro_use]` is it feels the same as when `std` is enabled and it's already in the prelude.